cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -138,4 +316,470 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        MetricType.COUNTER,
+        ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, streamId.getNamespaceId(),
+                        Constants.Metrics.Tag.STREAM, streamId.getName()),
+        ImmutableList.<String>of()
+      );
+
+      try {
+        Collection<MetricTimeSeries> metrics = metricStore.query(metricDataQuery);
+        if (metrics == null || metrics.isEmpty()) {
+          // Data is not yet available, which means no data has been ingested by the stream yet
+          return new StreamSize(0L, System.currentTimeMillis());
+        }
+
+        MetricTimeSeries metric = metrics.iterator().next();
+        List<TimeValue> timeValues = metric.getTimeValues();
+        if (timeValues == null || timeValues.size() != 1) {
+          throw new IOException("Should collect exactly one time value");
+        }
+        TimeValue timeValue = timeValues.get(0);
+        // The metric store gives us 0 as the timestamp, hence we cannot use it here
+        return new StreamSize(timeValue.getValue(), System.currentTimeMillis());
+      } catch (Exception e) {
+        Throwables.propagateIfInstanceOf(e, IOException.class);
+        throw new IOException(e);
+      }
+    }
+  }
+
+  /**
+   * Wrapper around a {@link StreamSizeSchedule} which will run a program whenever it receives enough
+   * data from a stream, via notifications.
+   */
+  private final class StreamSizeScheduleTask {
+    private final Id.Program programId;
+    private final SchedulableProgramType programType;
+    private StreamSizeSchedule streamSizeSchedule;
+
+    private long baseSize;
+    private long baseTs;
+    private AtomicBoolean active;
+
+    private StreamSizeScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule) {
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+    }
+
+    public void startSchedule(long baseSize, long baseTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with baseSize {}, baseTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), baseSize, baseTs, active, persist);
+      this.baseSize = baseSize;
+      this.baseTs = baseTs;
+      this.active = new AtomicBoolean(active);
+    }
+
+    public boolean isActive() {
+      return active.get();
+    }
+
+    public void received(StreamSizeNotification notification) {
+      if (!active.get()) {
+        return;
+      }
+      long pastRunSize;
+      long pastRunTs;
+      StreamSizeSchedule currentSchedule;
+      synchronized (this) {
+        if (notification.getSize() < baseSize) {
+          // This can happen if no notification is received for a stream for some time, and we poll the stream events
+          // size using metrics, and some metric events have hit their TTL. In that case it's impossible to know
```

Comment:
For the TTL issue, this is what I understood:
As long as at least one event is ingested by the stream in the period of the TTL, then the counter will never decrease. The counter could suddenly drop to 0 if no event has been ingested in TTL time. I'll double check with @abaranau.
This could lead to some tricky situations: if the stream already has 1.5 GB of data, and no data is ingested in TLL time, the metrics could suddenly show 0, yet the notification counting logic is unaware of that. If 0.5GB are ingested, the notification logic will send: 2GB, but the metrics will show 0.5GB. The polling and the notifications will never be in sync again, until CDAP is restarted and the notifications start counting from the size given by the metrics again.